File: src/main/java/jm/app/optimize/SteepestDescentOptimizer.java
package jm.app.optimize;

import jm.app.algebra.AlgebraUtil;
import jm.app.algebra.Matrix;

import java.math.BigDecimal;

public class SteepestDescentOptimizer implements Optimizer {

    private int epochNum = 100000;

    private BigDecimal learningRate = new BigDecimal(0.0001);

    public SteepestDescentOptimizer() {
    }

    public SteepestDescentOptimizer(int epochNum, double learningRate) {
        this.epochNum = epochNum;
        this.learningRate = new BigDecimal(learningRate);
    }

    @Override
    public Matrix optimize(TargetFunction targetFunction, Matrix params, Matrix args) {
        Matrix lastParams = AlgebraUtil.copy(params);
        Matrix newParams = AlgebraUtil.copy(params);
        for (int e = 0; e < epochNum; e++) {
            for (int i = 0; i < newParams.getColNum(); i++) {
                BigDecimal partialDerivative = calcPartialDerivative(targetFunction, lastParams, i, args);
                BigDecimal param = newParams.getValue(0, i).subtract(learningRate.multiply(partialDerivative));
                newParams.setValue(0, i, param);
            }
            lastParams = AlgebraUtil.copy(newParams);
        }
        return newParams;
    }

    @Override
    public Matrix optimize(TargetFunction targetFunction, Matrix params, Matrix trainInput, Matrix truthOutput, boolean toWrapRmseFunction) {
        if (toWrapRmseFunction) {
            targetFunction = new RmseFunction(targetFunction);
        }
        Matrix args = AlgebraUtil.mergeMatrix(trainInput, truthOutput, 1);
        Matrix lastParams = AlgebraUtil.copy(params);
        Matrix newParams = AlgebraUtil.copy(params);
        for (int e = 0; e < epochNum; e++) {
            for (int i = 0; i < newParams.getColNum(); i++) {
                BigDecimal partialDerivative = calcPartialDerivative(targetFunction, lastParams, i, args);
                BigDecimal param = newParams.getValue(0, i).subtract(learningRate.multiply(partialDerivative));
                newParams.setValue(0, i, param);
            }
            lastParams = AlgebraUtil.copy(newParams);
        }
        return newParams;
    }

    /**
     * Calculate the partial derivative with the target function, its corresponding parameters, and its arguments
     * @param targetFunction the target function that is guaranteed to be derivable
     * @param params the parameters that is represented by a matrix with only one row
     * @param partialVariableIndex the position index of the parameter that is to be calculated the partial derivative
     * @param args the arguments of the target function
     * @return the partial derivative
     */
    public BigDecimal calcPartialDerivative(TargetFunction targetFunction, Matrix params, int partialVariableIndex, Matrix args) {
        final BigDecimal epsilon = new BigDecimal(0.00000001);

        BigDecimal param = params.getValue(0, partialVariableIndex);
        BigDecimal leftOffset = param.subtract(epsilon);
        BigDecimal rightOffset = param.add(epsilon);
        Matrix leftOffsetParams = AlgebraUtil.copy(params);
        leftOffsetParams.setValue(0, partialVariableIndex, leftOffset);
        Matrix rightOffsetParams = AlgebraUtil.copy(params);
        rightOffsetParams.setValue(0, partialVariableIndex, rightOffset);

        BigDecimal fx = targetFunction.fun(params, args);
        BigDecimal leftDerivative = targetFunction.fun(leftOffsetParams, args).subtract(fx).multiply(new BigDecimal(1.0 / -epsilon.doubleValue()));
        BigDecimal rightDerivative = targetFunction.fun(rightOffsetParams, args).subtract(fx).multiply(new BigDecimal(1.0 / epsilon.doubleValue()));
        BigDecimal partialDerivative = leftDerivative.add(rightDerivative).multiply(new BigDecimal(0.5));
        return partialDerivative;
    }

    public int getEpochNum() {
        return epochNum;
    }

    public void setEpochNum(int epochNum) {
        this.epochNum = epochNum;
    }

    public BigDecimal getLearningRate() {
        return learningRate;
    }

    public void setLearningRate(BigDecimal learningRate) {
        this.learningRate = learningRate;
    }
}

File: py/optimize/steepest_descent_optimizer.py
import py.optimize.rmse_function as rf
import py.optimize.optimizer as opt
import numpy as np

class SteepestDescentOptimizer(opt.Optimizer):

    def __init__(self, epoch_num = 100000, learning_rate = 0.0001):
        self.epoch_num = epoch_num
        self.learning_rate = learning_rate

    def calc_partial_derivative(self, target_function, params, partial_variable_index, args):
        epsilon = 0.00000001
        param = params[partial_variable_index]
        left_offset = param - epsilon
        right_offset = param + epsilon
        left_offset_params = np.array(params)
        left_offset_params[partial_variable_index] = left_offset
        right_offset_params = np.array(params)
        right_offset_params[partial_variable_index] = right_offset
        fx = target_function.fun(params, args)
        left_derivative = np.divide(target_function.fun(left_offset_params, args) - fx, -epsilon)
        right_derivative = np.divide(target_function.fun(right_offset_params, args) - fx, epsilon)
        partial_derivative = np.multiply(left_derivative + right_derivative, 0.5)
        return partial_derivative

    def optimize(self, target_function, params, train_input, truth_output, to_wrap_rmse_function=True):
        if to_wrap_rmse_function == True:
            target_function = rf.RmseFunction(target_function)
        args = np.concatenate((train_input, truth_output), 1)
        last_params = np.array(params)
        new_params = np.array(params)
        for e in range(self.epoch_num):
            print("Epoch #%d" %e, new_params)
            for i in range(new_params.shape[0]):
                partial_derivative = self.calc_partial_derivative(target_function, last_params, i, args)
                param = np.subtract(new_params[i], self.learning_rate * partial_derivative)
                new_params[i] = param
            last_params = np.array(new_params)
        return new_params

